metron-platform/metron-enrichment/src/main/java/org/apache/metron/threatintel/triage/ThreatTriageProcessor.java

```diff
@@ -52,15 +74,36 @@ public ThreatTriageProcessor( SensorEnrichmentConfig config
 
   @Nullable
   @Override
-  public Double apply(@Nullable Map input) {
-    List<Number> scores = new ArrayList<>();
+  public ThreatScore apply(@Nullable Map input) {
+
+    ThreatScore threatScore = new ThreatScore();
     StellarPredicateProcessor predicateProcessor = new StellarPredicateProcessor();
+    StellarProcessor processor = new StellarProcessor();
     VariableResolver resolver = new MapVariableResolver(input, sensorConfig.getConfiguration(), threatIntelConfig.getConfig());
+
+    // attempt to apply each rule to the threat
     for(RiskLevelRule rule : threatTriageConfig.getRiskLevelRules()) {
       if(predicateProcessor.parse(rule.getRule(), resolver, functionResolver, context)) {
-        scores.add(rule.getScore());
+
+        // add the rule's score to the overall threat score
+        String reason = execute(rule.getReason(), processor, resolver, String.class);
+        RuleScore score = new RuleScore(rule, reason);
+        threatScore.addRuleScore(score);
       }
     }
-    return threatTriageConfig.getAggregator().aggregate(scores, threatTriageConfig.getAggregationConfig());
+
+    // calculate the aggregate threat score
+    Aggregators aggregators = threatTriageConfig.getAggregator();
+    List<Number> allScores = threatScore.getRuleScores().stream().map(score -> score.getRule().getScore()).collect(Collectors.toList());
+    Double aggregateScore = aggregators.aggregate(allScores, threatTriageConfig.getAggregationConfig());
+
+    // return the overall threat score
+    threatScore.setScore(aggregateScore);
+    return threatScore;
+  }
+
+  private <T> T execute(String expression, StellarProcessor processor, VariableResolver resolver, Class<T> clazz) {
+    Object result = processor.parse(expression, resolver, functionResolver, context);
```

Comment:
It is a cool idea @justinleet.  

If we could build the abstract syntax tree and then cache that, we could certainly improve performance.  Of course, we'd have to confirm how much time we spend building the AST versus execution, but I'd be willing to bet you are right.

I don't know that we have the ability to do this today though.  At least, I have not seen us do something like this elsewhere.  But I could be wrong.  

Might be a good general strategy to explore for improving performance in a lot of places considering how much we rely on Stellar executions nowadays.